### ISSUE TITLE
Deadlock caused by scheduling in callbacks may appear when I use kpatch

### ISSUE THREAD

**Issue by zhenjiangma** [dynup/kpatch#1120]
I found a bug in kpatch, which may cause deadlock. 
  I do some pre-patch preparation work in pre_patch_callback function, and in the preparation work the task named P1 will gave up CPU for some reason. At the same time, there is another task named P2 running which will send IPI.
  However, because the irq has been disabled in stop_machine function, and the P1 also goes to sleep due to pre_patch_callback function. Exactly, the P2 is scheduled. 
  Therefore, terrible things happen. P1 has disabled all irq and went to sleep, while P2 sends IPI and waits for responses. Inevitably the system trapped into the deadlock state. The reason for this is that the pre_patch/post_patch work is executed in stop_machine.  Is there a better way to avoid this scenario？
  In order to reproduce it, I specially designed the following scene.
/*** this is my patch file ***/
````diff --git a/fs/proc/meminfo.c b/fs/proc/meminfo.c
index 71d006c..6a93a52 100644
--- a/fs/proc/meminfo.c
+++ b/fs/proc/meminfo.c
@@ -22,6 +22,10 @@
 #include "internal.h"
 #include <linux/pid_namespace.h>
 
+#include <linux/module.h>
+#include "kpatch-macros.h"
+#include <linux/kernel.h>
+#include <linux/semaphore.h>
 void __attribute__((weak)) arch_report_meminfo(struct seq_file *m)
 {
 }
@@ -189,6 +193,39 @@ static int meminfo_proc_show(struct seq_file *m, void *v)
 
 	return 0;
 }
+static void callback_info(const char *callback, patch_object *obj)
+{   
+	    pr_info("ENTERING INTO CALLBACK_INFO");
+}
+
+static int pre_patch_callback(patch_object *obj)
+{    
+        struct semaphore sem;
+	    callback_info(__func__, obj);
+		sema_init(&sem, 1);
+	    down(&sem);
+		down(&sem);
+		return 0;
+}
+KPATCH_PRE_PATCH_CALLBACK(pre_patch_callback);
+
+static void post_patch_callback(patch_object *obj)
+{   
+	    callback_info(__func__, obj);
+}
+KPATCH_POST_PATCH_CALLBACK(post_patch_callback);
+
+static void pre_unpatch_callback(patch_object *obj)
+{
+	    callback_info(__func__, obj);
+}
+KPATCH_PRE_UNPATCH_CALLBACK(pre_unpatch_callback);
+
+static void post_unpatch_callback(patch_object *obj)
+{
+	    callback_info(__func__, obj);
+}
+KPATCH_POST_UNPATCH_CALLBACK(post_unpatch_callback);
 
 static int __init proc_meminfo_init(void)
 {
````

/*** this is my .ko intented to send IPI ***/
`````
#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/init.h>
#include <linux/smp.h>
#include <linux/sched.h>
#include <linux/kthread.h>
#include <linux/delay.h>
#include <linux/smp.h>
#include <linux/cpumask.h>
#include <linux/sched/types.h>

MODULE_LICENSE("GPL");
MODULE_AUTHOR("zhenjiang.mzj");
MODULE_DESCRIPTION("deadlock test");
MODULE_VERSION("0.1");

static struct task_struct *my_task = NULL;

static void ipi_send(void * info){
//	pr_info("sending ipi ...");
}	

static int my_kthread(void *data) {
    char *str = (char *)data;
    pr_info("my kthread data: %s\n", str);
    pr_info("my kthread smp_processor_id %d\n", smp_processor_id());
    while(!kthread_should_stop()) {
        // pr_info("prepare to send IPI to all processors.");
        **smp_call_function_many(cpu_online_mask, ipi_send, NULL, true);**
	schedule();
        // pr_info("send IPI done.\n");
    }
    return 0;
}

static int __init my_init(void)
{
    struct sched_param param = {.sched_priority= 99};
    pr_info("my_init is starting\n");
    pr_info("smp_processor_id %d\n", smp_processor_id());
    my_task = kthread_run(my_kthread, "my kthread is starting", "mykthread-%s", "test");
    sched_setscheduler(my_task, SCHED_FIFO, &param);
    pr_info("my init finish.\n");
    pr_info("=========================================\n");
    return 0;
}

static void __exit my_exit(void)
{
    pr_info("my exit.\n");
    pr_info("smp_processor_id %d\n", smp_processor_id());
    if (my_task) {
        pr_info("stop kthread\n");
        kthread_stop(my_task);
    }
    pr_info("my exit finish.\n");
}

module_init(my_init);
module_exit(my_exit);
`````

**Comment by julien-thierry** [comment]
Hi zhenjiangma,

I might be missing something obvious. But isn't the deadlock caused by the use of the semaphore in your pre-patch callback?

    sema_init(&sem, 1);

Means that the semaphore only allows one taker at a time. And then `down(&sem)` is called twice in the same function. Without any calls to `up(&sem)`. My understanding is that this will deadlock the task performing the pre-patch.

**Comment by zhenjiangma** [comment]
> Hi zhenjiangma,
> 
> I might be missing something obvious. But isn't the deadlock caused by the use of the semaphore in your pre-patch callback?
> 
> ```
> sema_init(&sem, 1);
> ```
> 
> Means that the semaphore only allows one taker at a time. And then `down(&sem)` is called twice in the same function. Without any calls to `up(&sem)`. My understanding is that this will deadlock the task performing the pre-patch.

In order to reproduce this phenomenon, twice '''down(&sem)''' were executed. In fact, the problem is not what I performed. Even if I do not perform this kind of operation, other situations that may cause sleep/switch task during executing callbacks will also cause this problem.

**Comment by xiejingf** [comment]
I  have encountered the same issue before,  as mentioned by #1077

**Comment by jpoimboe** [comment]
As mentioned in #1077, kpatch.ko is deprecated, but pull requests are welcome.

**Comment by joe-lawrence** [comment]
Fixed by PR #1134 ("deal with the deadlock may occur in stop_machine context")
